itests/gateway_test.go

```diff
@@ -297,3 +359,209 @@ func sendFunds(ctx context.Context, fromNode *kit.TestFullNode, fromAddr address
+		t.Logf("Received subscription response (sub1): %v", resp)
+		return nil
+	})
+	req.NoError(err)
+	subId2, err := nodes.lite.EthSubscribe(ctx, res.Wrap[jsonrpc.RawParams](json.Marshal(ethtypes.EthSubscribeParams{EventType: "newHeads"})).Assert(req.NoError))
+	req.NoError(err)
+	err = nodes.lite.EthSubRouter.AddSub(ctx, subId2, func(ctx context.Context, resp *ethtypes.EthSubscriptionResponse) error {
+		t.Logf("Received subscription response (sub2): %v", resp)
+		return nil
+	})
+	req.NoError(err)
+
+	ok, err := nodes.lite.EthUnsubscribe(ctx, subId1) // unsub on lite node, should work
+	req.NoError(err)
+	req.True(ok)
+	ok, err = nodes.full.EthUnsubscribe(ctx, subId1) // unsub on full node, already done
+	req.NoError(err)
+	req.False(ok)
+
+	t.Logf("Installing a stateful filters via ws")
+	// install the variety of stateful filters we have, but only up to the max total
+	var (
+		blockFilterIds   = make([]ethtypes.EthFilterID, gateway.DefaultEthMaxFiltersPerConn/3)
+		pendingFilterIds = make([]ethtypes.EthFilterID, gateway.DefaultEthMaxFiltersPerConn/3)
+		// matchFilterIds takes up the remainder, minus 1 because we still have 1 live subscription that counts
+		matchFilterIds = make([]ethtypes.EthFilterID, gateway.DefaultEthMaxFiltersPerConn-len(blockFilterIds)-len(pendingFilterIds)-1)
+	)
+	for i := 0; i < len(blockFilterIds); i++ {
+		fid, err := nodes.lite.EthNewBlockFilter(ctx)
+		req.NoError(err)
+		blockFilterIds[i] = fid
+	}
+	for i := 0; i < len(pendingFilterIds); i++ {
+		fid, err := nodes.lite.EthNewPendingTransactionFilter(ctx)
+		req.NoError(err)
+		pendingFilterIds[i] = fid
+	}
+	for i := 0; i < len(matchFilterIds); i++ {
+		fid, err := nodes.lite.EthNewFilter(ctx, &ethtypes.EthFilterSpec{})
+		req.NoError(err)
+		matchFilterIds[i] = fid
+	}
+
+	// sanity check we're actually doing something
+	req.Greater(len(blockFilterIds), 0)
+	req.Greater(len(pendingFilterIds), 0)
+	req.Greater(len(matchFilterIds), 0)
+
+	t.Logf("Testing 'too many filters' rejection")
+	_, err = nodes.lite.EthNewBlockFilter(ctx)
+	require.ErrorContains(t, err, "too many subscriptions and filters for this connection")
+	_, err = nodes.lite.EthNewPendingTransactionFilter(ctx)
+	require.ErrorContains(t, err, "too many subscriptions and filters for this connection")
+	_, err = nodes.lite.EthNewFilter(ctx, &ethtypes.EthFilterSpec{})
+	require.ErrorContains(t, err, "too many subscriptions and filters for this connection")
+	_, err = nodes.lite.EthSubscribe(ctx, res.Wrap[jsonrpc.RawParams](json.Marshal(ethtypes.EthSubscribeParams{EventType: "newHeads"})).Assert(req.NoError))
+	require.ErrorContains(t, err, "too many subscriptions and filters for this connection")
+
+	t.Logf("Shutting down the lite node")
+	req.NoError(nodes.lite.Stop(ctx))
+	nodes.rpcCloser() // once the websocke connection is closed, the server should clean up the filters for us
+
+	time.Sleep(time.Second) // unfortunately we have no other way to check for completeness of shutdown and cleanup
```

Comment:
Can you explain how the sleep here assures completeness of shutdown and cleanup?

Reply:
expanded on the comment in here:

```
	// Once the websocket connection is closed, the server should clean up the filters for us.
	// Unfortunately we have no other way to check for completeness of shutdown and cleanup.
	// * Asynchronously the rpcCloser() call will end the client websockets connection to the gateway.
	// * When the gateway recognises the end of the HTTP conneciton, it will asynchronously make calls
	//   to the fullnode to clean up the filters.
	// * The fullnode will then uninstall the filters and we can finally move on to check it directly
	//   that this has happened.
	// This should happen quickly, but we have no way to synchronously check for it. So we just wait a bit.
```

i.e. multiple layers of async over 2 separate HTTP connections and no direct insight to internal state of either gateway or fullnode. I also can't think of a good "eventually" that's idempotent enough to use here.

Reply:
I see the code is full of time.Sleeps from priror work. Avoiding those seems like a separate body of work.

Thank you for the comment; that helps me as a reader a lot 👍